kong/plugins/http-log/handler.lua

```diff
@@ -71,67 +96,50 @@ local function send_payload(self, conf, payload)
   local parsed_url = parse_url(http_endpoint)
   local host = parsed_url.host
   local port = tonumber(parsed_url.port)
+  local userinfo = parsed_url.userinfo
 
   local httpc = http.new()
   httpc:set_timeout(timeout)
 
-  table_clear(headers_cache)
+  local headers = {
+    ["Host"] = host,
+    ["Content-Type"] = content_type,
+    ["Content-Length"] = content_length,
+    ["Authorization"] = userinfo and "Basic " .. encode_base64(userinfo) or nil
+  }
   if conf.headers then
     for h, v in pairs(conf.headers) do
-      headers_cache[h] = v
+      headers[h] = headers[h] or v -- don't override Host, Content-Type, Content-Length, Authorization
     end
   end
 
-  headers_cache["Host"] = parsed_url.host
-  headers_cache["Content-Type"] = content_type
-  headers_cache["Content-Length"] = #payload
-  if parsed_url.userinfo then
-    headers_cache["Authorization"] = "Basic " .. encode_base64(parsed_url.userinfo)
-  end
+  local upstream_url = fmt("%s://%s:%d%s", parsed_url.scheme, host, port, parsed_url.path)
 
-  params_cache.method = method
-  params_cache.body = payload
-  params_cache.keepalive_timeout = keepalive
-
-  local url = fmt("%s://%s:%d%s", parsed_url.scheme, parsed_url.host, parsed_url.port, parsed_url.path)
-
-  -- note: `httpc:request` makes a deep copy of `params_cache`, so it will be
-  -- fine to reuse the table here
-  local res, err = httpc:request_uri(url, params_cache)
+  local res, err = httpc:request_uri(upstream_url, {
+    method = method,
+    headers = headers,
+    body = payload,
```

Comment:
here `payload` just a function?

Reply:
Yes.  It will be called by the http client to yield parts of the body until it returns `nil`.  It is used here to avoid having to copy the payload into a temporary string for sending.